scheduler/scripts/run_with_events.py

```diff
@@ -0,0 +1,131 @@
+                                                             sb=SkyBackground.SBANY,
+                                                             wv=WaterVapor.WVANY),
+                                   start=datetime(2018, 10, 1, 10),
+                                   reason='Worst image quality',
+                                   )
+    queue.add_events(weather_change, 0)
+
+    builder = ValidationBuilder(Sources(), queue)
+
+    start = Time("2018-10-01 08:00:00", format='iso', scale='utc')
+    end = Time("2018-10-03 08:00:00", format='iso', scale='utc')
+    num_nights_to_schedule = int(round(end.jd - start.jd)) + 1
+    collector = builder.build_collector(
+        start=start,
+        end=end,
+        sites=ALL_SITES,
+        semesters=frozenset([Semester(2018, SemesterHalf.B)]),
+        blueprint=collector_blueprint
+    )
+    # Create the Collector and load the programs.
+    collector.load_programs(program_provider_class=OcsProgramProvider,
+                            data=programs)
+
+    ValidationBuilder.update_collector(collector)  # ZeroTime observations
+
+    # TODO: SET THE WEATHER HERE.
+    # CC values are CC50, CC70, CC85, CCANY. Default is CC50 if not passed to build_selector.
+    cc = CloudCover.CC50
+
+    # IQ values are IQ20, IQ70, IQ85, and IQANY. Default is IQ70 if not passed to build_selector.
+    iq = ImageQuality.IQ70
+
+    selector = builder.build_selector(collector,
+                                      num_nights_to_schedule=num_nights_to_schedule,
+                                      default_cc=cc,
+                                      default_iq=iq)
+
+    # Prepare the optimizer.
+    optimizer_blueprint = OptimizerBlueprint(
+        "GreedyMax"
+    )
+    optimizer = builder.build_optimizer(
+        blueprint=optimizer_blueprint
+    )
+
+    # The total nights for which visibility calculations have been done.
+    total_nights = len(collector.time_grid)
+
+    # Create the overall plans by night.
+    overall_plans = {}
+    night_timeline = NightTimeline({})
+    for night_idx in range(selector.num_nights_to_schedule):
+
+        events_by_night = queue.get_night_events(night_idx)
+        night_indices = np.array([night_idx])
+        # changes = NightChanges()
+
+        # Run eventless timeline
+        selection = selector.select(night_indices=night_indices)
+        # Run the optimizer to get the plans for the first night in the selection.
+        plans = optimizer.schedule(selection)
+        night_timeline.add(night_idx, 0, Twilight(start.to_datetime()), plans[0])
+
+        if events_by_night:
+            while events_by_night:
+                event = events_by_night.pop()
+                event_start_time_slot = ceil((event.start - start.to_datetime()).total_seconds()/60)
+                if isinstance(event, WeatherChange):
+                    selector.default_iq = event.new_conditions.iq
+                    selector.default_cc = event.new_conditions.cc
```

Comment:
Good, and okay for now, but I think we need to start thinking of moving all of this logic out of a script and into an object.